### DISCUSSION TITLE
Relation Manager Action URL always running without click on listing

### DISCUSSION THREAD

**Discussion by AFSDARIF** [filamentphp/filament#6911] [Help] [answered]
![image](https://github.com/filamentphp/filament/assets/10229660/dbd02888-8c4a-4719-bea4-57a25af79f52)

Why a url function is alway run in reload listing parent of Relation Manager?

**Comment by pboivin** [comment]
I think it's because the component gets re-rendered... Using a callback like you are doing (`->url(function () { ... })`) gives you an opportunity to change the value dynamically.

**Reply by AFSDARIF** [reply]
ah okay. For my case, how to make a custom action without re-rendered?

**Reply by pboivin** [reply]
I'm missing a bit of context for your project so this may not be the best solution, but if you need to pass a closure to `->url()` and only run it once, you could probably add a public property on your RelationManager component:

```php
public string $receiptActionUrl = '';
```

and then:
```php
    // ...
    
    ->url(function ($record, $livewire) {
        if (!$livewire->receptActionUrl) {
            $livewire->receptActionUrl = /* generate url */
        }
        return $livewire->receptActionUrl;
    })
```

**Reply by AFSDARIF** [reply, marked as answer]
ah yes, I think I'm also wrong for not providing enough information about the purpose of what I'm going to do. in the code that I attached, I intend to generate receipts from the PDF API that I created, therefore I need to run the generate function.

I haven't tried your suggestion yet, but I have found an alternative.
```
Tables\Actions\Action::make('generate_receipt')
                    ->label('Download Receipt')->icon('heroicon-o-download')
                    ->hidden(function($record, RelationManager $livewire) {
                        if($livewire->getRelationship()->getParent()->receipt_order_provider == null || $livewire->getRelationship()->getParent()->receipt_order_provider == ''){
                            return true;
                        }else{
                            return false;
                        }
                    })
                    ->action('generateReceipt'),
```

and then download a file PDF
```
public function generateReceipt($record, RelationManager $livewire)
    {
      // ...
      return response()->streamDownload(function() use ($path,$record) {
            echo file_get_contents(Storage::disk('oss')->url($path));
        }, $record->bill_code.'.pdf');
    }
```
